File: Linked Lists/SinglyLinkedList.py
from Node import Node

class SinglyLinkedList:

    def __init__(self) -> None:
        self.head = None
        self.size = 0

    def append(self, val):

        # OBJECTIVE: Add a new node to the end of the list

        # Create a new node
        newNode = Node()
        newNode.val = val

        # If head is by itself, add newNode as head
        if self.head == None:
            self.head = newNode
        
        # If not, add newNode before tail node
        else:

            # Traverse linked list to the end
            curNode = self.head
            while curNode.next != None:
                curNode = curNode.next

            # Link last node with new node
            curNode.next = newNode

        self.size += 1

    def getHead(self):

        # OBJECTIVE: Return head node if it exists

        # If there's no head node, exit function
        if self.head == None:
            return

        return self.head

    def getSize(self):

        # OBJECTIVE: Get linked list size
        return self.size

    def printList(self):

        # OBJECTIVE: Print linked list from head

        # If linked list is empty, exit function
        if self.head == None:
            print("Linked list is empty")
            return

        # Get head node
        curNode = self.head

        # Traverse linked list
        while curNode != None:
            print(curNode.val, end=" ")
            curNode = curNode.next

File: Linked_Lists/SinglyLinkedList.java

public class SinglyLinkedList {

    Node headNode = null;
    int size = 0;

    void append(int newVal) {

        // OBJECTIVE: Add a new node to the end of the list

        // Create a new node
        Node newNode = new Node(newVal);

        // If linked list is empty, add newNode as head
        if (headNode == null) {
            headNode = newNode;
        }
        else {

            // Traverse linked list
            Node curNode = headNode;
            while (curNode.next != null) {
                curNode = curNode.next;
            }

            // Connect last node to newNode
            curNode.next = newNode;
        }

        size++;
    }

    Node getHead() {

        // OBJECTIVE: Return node after headNode
        
        // If there's no node after headNode, exit function
        if (headNode == null) {return null;}

        return headNode;
    }

    Node getMiddle() {

        // OBJECTIVE: Return middle node from linked list

        // If linked list is empty, exit function
        if (headNode == null) {return null;}

        // Create a temporary node
        Node tmpNode = headNode;

        // Move to node until midpoint
        for (int i=0; i<size/2; i++) {
            tmpNode = tmpNode.next;
        }

        return tmpNode;
    }

    int getSize() {

        // OBJECTIVE: Return linked list size
        return size;
    }

    void printList() {

        // OBJECTIVE: Print linked list

        // If linked list is empty, exit function
        if (headNode == null) {
            System.out.println("Linked list is empty!");
            return;
        }

        // Get headNode
        Node curNode = headNode;

        // Traverse linked list
        while (curNode != null) {
            System.out.print(curNode.val + " ");
            curNode = curNode.next;
        }
    }
}